storage: Support deferred cancellation in the CommandQueue

nvb opened the pull request:
Fixes #9237.

Previously when a command was cancelled while in the `CommandQueue` due to
a `Context` cancellation, we would spawn a goroutine to block until all
prerequisite commands completed. This change allows us to avoid the
extra goroutine by offloading the cleanup work to the abandoned commands'
dependents.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/9448)

<!-- Reviewable:end -->

nvb committed:
storage: Support deferred cancellation in the CommandQueue

Fixes #9237.

Previously when a command was cancelled while in the CommandQueue due to
a Context cancellation, we would spawn a goroutine to block until all
prerequisite commands completed. This change allows us to avoid the
extra goroutine by offloading the cleanup work to the abandoned commands'
dependents.

nvanbenschoten commented:
This isn't actually working correctly right now, so don't bother reviewing yet.

nvanbenschoten commented:
Ok, this should be working again. I added much more comprehensive testing to assure that cancellation is correctly handled in various dependency configurations.

petermattis commented:
The idea seems sounds, though there is a lot of complexity in `Replica.beginCmds`. I'm going to have to go over that more thoroughly later. Targeting this PR at `develop` feels like a recipe for merge headaches. And I'm not sure if we should target it at `master` given that the current cancellation works. That seems to leave this PR in limbo until we exit code yellow.

nvanbenschoten commented:
Yeah there seems to be too much code movement here for it not to cause merge headaches on `develop`. I think we should target this at `master`, but I dont think it's a justifiable change to `master` during the code yellow.

bdarnell commented on util/syncutil/once.go:
```diff
@@ -0,0 +1,76 @@
+// that it holds a reference to a Mutex instead of holding the Mutex by value.
+// This means that it can be used in conjunction with external synchronization
+// that uses the same Mutex.
+type Once struct {
```
The types in `syncutil` that have the same names as ones from `sync` should be drop-in replacements. This isn't, because it requires the mutex to be initialized explicitly. 

Why does it matter that the lock in `Once` be the same as an external lock? Wouldn't it be the same to use a normal `sync.Once` and then acquire the external lock inside the callback?

bdarnell commented on util/syncutil/once.go:
```diff
@@ -0,0 +1,76 @@
+}
+
+// SetMutex sets the mutex that the Once object should use for synchronization.
+func (o *Once) SetMutex(m *Mutex) {
```
Why `*Mutex` instead of `sync.Locker`?

bdarnell commented on util/syncutil/once.go:
```diff
@@ -0,0 +1,76 @@
+	return o.IsDoneLocked()
+}
+
+// Avoid unused go vet warning. The method is useful enough to justify its existence.
```
Why? `sync.Once` doesn't have it (and this isn't just Go minimalism; similar mechanisms like `pthread_once` are the same). The idea behind `Once` is that you can always call `Do()` (which is as fast as `IsDone` once it's been done) and you never have to use any conditionals.

bdarnell commented on storage/replica.go:
```diff
@@ -1150,42 +1150,106 @@ func (r *Replica) beginCmds(ctx context.Context, ba *roachpb.BatchRequest) (func
+
+							// Our prerequisite cancelled before us and told each of its
+							// prerequisites that we were one of their new dependent commands.
+							// Decrement their dependents count and recursively assure that
```
This only decrements if newCmd.dependents is zero. Is that always the case here?

bdarnell commented on storage/replica.go:
```diff
@@ -1150,42 +1150,106 @@ func (r *Replica) beginCmds(ctx context.Context, ba *roachpb.BatchRequest) (func
+					return nil, err
+				}
+
+				// If the command does have dependents, we allow one of them to clean it
```
I don't understand why it's not safe to remove the command from the queue here but it is safe for the first dependent to do so (without waiting for all the other dependents to resolve).

bdarnell reviewed (commented):
Yeah, for the remainder of the code yellow we'd rather not make changes to core packages on the `develop` branch; this should be left unmerged until we're ready to put it on master.

nvb committed:
storage: Pull the addition of a prereq in getPrereqs into helper closure

nvanbenschoten commented:
Thanks for the review Ben. Now that we're getting close to exiting the code yellow, I figured I'd revisit this change.

---

Review status: 0 of 4 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[storage/command_queue.go, line 42 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-KRvHMk1k55Qg83Wq8t2-r3-42:-KSlbl_vyBP4KYQSO90X:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c515626569cd0ea42fc82f493ba2e111acebf439/storage/command_queue.go#L42)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> s/commands'/command's/
</details>
Done.

---

*[storage/replica.go, line 1240 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-KRvHMk1k55Qg83Wq8t4-r3-1240:-KSogIGLG2hb-pH5nsGm:bm8pmys) ([raw file](https://github.com/cockroachdb/cockroach/blob/c515626569cd0ea42fc82f493ba2e111acebf439/storage/replica.go#L1240)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> I don't understand why it's not safe to remove the command from the queue here but it is safe for the first dependent to do so (without waiting for all the other dependents to resolve). 
</details>
When initially writing this I was under the assumption that there was a good reason for this restriction, but after rethinking it I don't see why it wouldn't be safe to remove commands immediately after cancellation from the tree. I've made the change, which allowed a lot of this code to be simplified, and allowed the entire first commit to be removed.

---

*[util/syncutil/once.go, line 27 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-KRvHMk-JSVu5WA-8v5E-r3-27:-KSoh6sdi_xW_XIlQj_o:bssvfic) ([raw file](https://github.com/cockroachdb/cockroach/blob/c515626569cd0ea42fc82f493ba2e111acebf439/util/syncutil/once.go#L27)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> The types in `syncutil` that have the same names as ones from `sync` should be drop-in replacements. This isn't, because it requires the mutex to be initialized explicitly. 
> 
> 
> 
> Why does it matter that the lock in `Once` be the same as an external lock? Wouldn't it be the same to use a normal `sync.Once` and then acquire the external lock inside the callback?
</details>
The idea was to avoid acquiring multiple locks by using the same lock for the slow-path of the `Once` object as the lock needed to perform the action in the `Do` closure. Depending on the contention on the external lock, this may or may not have a performance benefit because it avoids nested synchronization.

---

*[util/syncutil/once.go, line 69 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-KRvHMk-JSVu5WA-8v5E-r3-69:-KSoiK1QEx6tm6uGczoX:bphklwr) ([raw file](https://github.com/cockroachdb/cockroach/blob/c515626569cd0ea42fc82f493ba2e111acebf439/util/syncutil/once.go#L69)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Why? `sync.Once` doesn't have it (and this isn't just Go minimalism; similar mechanisms like `pthread_once` are the same). The idea behind `Once` is that you can always call `Do()` (which is as fast as `IsDone` once it's been done) and you never have to use any conditionals. 
</details>
I've run into a few situations where I'd like to check if a `Once` object has completed its work in a safe way. The example that's coming to mind is the old `wasEvicted` method on the `evictionToken` in `range_cache.go`. I don't see any harm in adding a method like this, especially because it's surprisingly tricky to safely maintain/check this state using some form of external synchronization. That said, maybe the desire for this method shows that I was trying to use the wrong tool for the job.

Even though I'm no longer proposing this commit, I'm curious why we wouldn't want this added functionality.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-:-KSom6INSf9cQ5gMVfH7:bmek2hm)*
<!-- Sent from Reviewable.io -->

tamird commented:
mostly <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>, nice work!

---

Reviewed 1 of 1 files at r1, 1 of 4 files at r2, 4 of 4 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 24 unresolved discussions, all commit checks successful.

---

*[storage/command_queue.go, line 43 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-KRwpfHvTRDFxp3M17pG:-KRwpfHvTRDFxp3M17pH:btvapr2) ([raw file](https://github.com/cockroachdb/cockroach/blob/494f6acdc2a5e281cfd4b38b756ee04cfb46eef4/storage/command_queue.go#L43)):*
> ```Go
> // After determining prerequisite commands and adding the new command to the
> // command queue, the new command should wait on each prerequisite commands'
> // pending channel for confirmation that all overlapping, pending commands have
> ```

s/, pending commands/commands/

---

*[storage/command_queue.go, line 33 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-KSr4lnCLVADM2Qwwpp-:-KSr4lnCLVADM2Qwwpp0:b-s3osk3) ([raw file](https://github.com/cockroachdb/cockroach/blob/695f5a1f988015678931bcc5291fbf3a9107d78d/storage/command_queue.go#L33)):*
> ```Go
> // Before executing, a command invokes getPrereqs() to acquire a slice of
> // references to overlapping commands that are already in the command queue. After
> // determining its prerequisite commands, a new command is added to the queue via add().
> ```

s/a new/the/

---

*[storage/command_queue.go, line 38 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-KSr4skiURFaqZYITGrg:-KSr4skiURFaqZYITGrh:b-8ye4ui) ([raw file](https://github.com/cockroachdb/cockroach/blob/695f5a1f988015678931bcc5291fbf3a9107d78d/storage/command_queue.go#L38)):*
> ```Go
> // returned via getPrereqs() don't include read-only on read-only overlapping commands
> // as an optimization. Both getPrereqs() and add() must see an atomic view of the command
> // queue, so in a concurrent setting, their execution must be synchronized.
> ```

they must also be done under the same lock.

---

*[storage/command_queue.go, line 105 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-KSr57jaSsayCuzxYyyX:-KSr57jaSsayCuzxYyyY:b-txw0wu) ([raw file](https://github.com/cockroachdb/cockroach/blob/695f5a1f988015678931bcc5291fbf3a9107d78d/storage/command_queue.go#L105)):*
> ```Go
> 
> // getPrereqs returns a slice of the prerequisite commands which overlap the
> // specified key ranges. If an end key is empty, it only affects the start key.
> ```

while you're here, "it" is ambiguous.

---

*[storage/command_queue.go, line 106 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-KSr5Dva9Kp4kKTNShdq:-KSr5Dva9Kp4kKTNShdr:b-auf6n7) ([raw file](https://github.com/cockroachdb/cockroach/blob/695f5a1f988015678931bcc5291fbf3a9107d78d/storage/command_queue.go#L106)):*
> ```Go
> // getPrereqs returns a slice of the prerequisite commands which overlap the
> // specified key ranges. If an end key is empty, it only affects the start key.
> // The caller should call add() to add the keys to the command queue and then
> ```

add() adds the command, not the keys, right? this comment seems malformed. simply, i would remove the mentions of "add" entirely, they're covered by the god comment above.

---

*[storage/command_queue.go, line 184 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-KSr5i6IZ0fPqT27YEPy:-KSr5i6IZ0fPqT27YEPz:btcrop7) ([raw file](https://github.com/cockroachdb/cockroach/blob/695f5a1f988015678931bcc5291fbf3a9107d78d/storage/command_queue.go#L184)):*
> ```Go
> 		// the new commands overlapped range.
> 		//
> 		// This approach can be improved further by noting that dependencies on overlapping
> ```

can you add the word TODO somewhere? otherwise it's difficult to discern if this is done already or not

---

*[storage/command_queue.go, line 338 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-KSr63MopGECURxV4It1:-KSr63MopGECURxV4It2:brndzgp) ([raw file](https://github.com/cockroachdb/cockroach/blob/695f5a1f988015678931bcc5291fbf3a9107d78d/storage/command_queue.go#L338)):*
> ```Go
> }
> 
> // overlapHeap is a max-heap of cmd references, sorting the elements
> ```

just "overlapHeap is a max-heap ordered by cmd.id." anything more is asking for maintenance or rot

---

*[storage/command_queue.go, line 378 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-KSr6cRuUePmOTBTvB8t:-KSr6cRuUePmOTBTvB8u:b-st34fn) ([raw file](https://github.com/cockroachdb/cockroach/blob/695f5a1f988015678931bcc5291fbf3a9107d78d/storage/command_queue.go#L378)):*
> ```Go
> //
> // add should be invoked before waiting on already-executing, overlapping
> // commands via the pending channels of prerequisites returned from getPrereqs().
> ```

can't you just refer to the `prereqs` argument here, instead of this verbose explanation?

"add should be invoked before waiting on prereqs' pending channels."

---

*[storage/command_queue.go, line 444 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-KSr6q4G_rau-CTm9Xtj:-KSr6q4G_rau-CTm9Xtk:b-qghjca) ([raw file](https://github.com/cockroachdb/cockroach/blob/695f5a1f988015678931bcc5291fbf3a9107d78d/storage/command_queue.go#L444)):*
> ```Go
> 
> func (cq *CommandQueue) removeFromTree(cmd *cmd) {
> 	if cmd == nil {
> ```

this can happen? seems sketchy

---

*[storage/command_queue.go, line 471 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-KSr6mxJF8EBM9acI_-n:-KSr6mxJF8EBM9acI_-o:b-qghjca) ([raw file](https://github.com/cockroachdb/cockroach/blob/695f5a1f988015678931bcc5291fbf3a9107d78d/storage/command_queue.go#L471)):*
> ```Go
> 
> func (cq *CommandQueue) closePendingChannels(cmd *cmd) {
> 	if cmd == nil {
> ```

this can happen? seems sketchy

---

*[storage/replica.go, line 1199 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-KSr8ZgET17INVsg2BQO:-KSr8ZgET17INVsg2BQP:b99evoa) ([raw file](https://github.com/cockroachdb/cockroach/blob/695f5a1f988015678931bcc5291fbf3a9107d78d/storage/replica.go#L1199)):*
> ```Go
> 					//  cmd 2:           -----
> 					//                     |
> 					//  cmd 3:     xxxxxxxxxxx
> ```

love this diagram

---

*[storage/replica.go, line 1211 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-KSr7UORwtD7ESdHyYzC:-KSr7UORwtD7ESdHyYzD:b-hgz9lb) ([raw file](https://github.com/cockroachdb/cockroach/blob/695f5a1f988015678931bcc5291fbf3a9107d78d/storage/replica.go#L1211)):*
> ```Go
> 				log.ErrEvent(ctx, errStr)
> 
> 				r.mu.Lock()
> ```

add a comment explaining why so much of this needs to be synchronized. also, i don't think this defer buys readability, i'd just move the call down.

---

*[storage/replica.go, line 1213 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-KSr8HnumoClKJkaSeBR:-KSr8HnumoClKJkaSeBS:b-4hpczu) ([raw file](https://github.com/cockroachdb/cockroach/blob/695f5a1f988015678931bcc5291fbf3a9107d78d/storage/replica.go#L1213)):*
> ```Go
> 				r.mu.Lock()
> 				defer r.mu.Unlock()
> 				newCmd.cancelled = true
> ```

this doesn't need to be under lock, right? in fact, it seems that only the command queue stuff needs to be under lock.

---

*[storage/replica.go, line 1215 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-KSr7iHw-Jf_313SbmOH:-KSr7iHxdx6CACluO4Jg:bvqjcya) ([raw file](https://github.com/cockroachdb/cockroach/blob/695f5a1f988015678931bcc5291fbf3a9107d78d/storage/replica.go#L1215)):*
> ```Go
> 				newCmd.cancelled = true
> 
> 				// Truncate the cancelled commands' prerequisite list to only those
> ```

command's

---

*[storage/replica.go, line 1216 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-KSr7pUE_MEhQrnKfLi6:-KSr7pUE_MEhQrnKfLi7:bj3gtem) ([raw file](https://github.com/cockroachdb/cockroach/blob/695f5a1f988015678931bcc5291fbf3a9107d78d/storage/replica.go#L1216)):*
> ```Go
> 
> 				// Truncate the cancelled commands' prerequisite list to only those
> 				// prerequisites that have not yet been waited on.
> ```

explain why you need to truncate the capacity?

---

*[storage/replica_test.go, line 1992 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-KSr8w8qySmd7HnHygXL:-KSr8w8qySmd7HnHygXM:bduk27r) ([raw file](https://github.com/cockroachdb/cockroach/blob/695f5a1f988015678931bcc5291fbf3a9107d78d/storage/replica_test.go#L1992)):*
> ```Go
> 	initial := commandInstruction{
> 		span: roachpb.Span{
> 			Key:    roachpb.Key("A"),
> ```

why not use keymin and keymax for minimum arbitrariness?

---

*[storage/replica_test.go, line 2036 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-KSr98LM-hddvR0fUeoT:-KSr98LM-hddvR0fUeoU:bvsehee) ([raw file](https://github.com/cockroachdb/cockroach/blob/695f5a1f988015678931bcc5291fbf3a9107d78d/storage/replica_test.go#L2036)):*
> ```Go
> 	}
> 
> 	// Cancel all commands that should be cancelled in a non-deterministic order.
> ```

wouldn't it be easier to use a set?

---

*[storage/replica_test.go, line 2068 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-KSr9JvYktStZpfq3ePG:-KSr9JvZnsJU0yaFCccM:b-nqsfa7) ([raw file](https://github.com/cockroachdb/cockroach/blob/695f5a1f988015678931bcc5291fbf3a9107d78d/storage/replica_test.go#L2068)):*
> ```Go
> 		t.Fatalf("expected command queue to be empty, found %d remaining commands", l)
> 	}
> 	tc.rng.mu.Unlock()
> ```

unlock this before `fatal`ing to avoid deadlocks during shutdown.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-:-KSr9m1Cgr2xWcULrZui:bvrcobc)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
LGTM, this is a nice simplification from before. I'd still like to wait for the dust to settle from the master/develop merge before this goes in, though.

---

Reviewed 1 of 4 files at r2, 2 of 4 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 21 unresolved discussions, all commit checks successful.

---

*[storage/replica_test.go, line 2036 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-KSr98LM-hddvR0fUeoT:-KStaC1w2gW1olRhG_O0:b3nj427) ([raw file](https://github.com/cockroachdb/cockroach/blob/695f5a1f988015678931bcc5291fbf3a9107d78d/storage/replica_test.go#L2036)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> wouldn't it be easier to use a set?
</details>
Or `rand.Perm`

---

*[util/syncutil/once.go, line 69 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-KRvHMk-JSVu5WA-8v5E-r3-69:-KStUscHwhjSlwlMpsZB:b-u84rdl) ([raw file](https://github.com/cockroachdb/cockroach/blob/c515626569cd0ea42fc82f493ba2e111acebf439/util/syncutil/once.go#L69)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> I've run into a few situations where I'd like to check if a `Once` object has completed its work in a safe way. The example that's coming to mind is the old `wasEvicted` method on the `evictionToken` in `range_cache.go`. I don't see any harm in adding a method like this, especially because it's surprisingly tricky to safely maintain/check this state using some form of external synchronization. That said, maybe the desire for this method shows that I was trying to use the wrong tool for the job.
> 
> Even though I'm no longer proposing this commit, I'm curious why we wouldn't want this added functionality.
</details>
My biggest objection was just to the fact that this was not a drop-in replacement for `sync.Once`. I don't have a clear objection to adding an `IsDone` method except to observe that other similar libraries have not found a persuasive need for its addition. 

The prototypical use for `Once` is for lazy initialization of a global; creating lots of Onces and asking about their status is unconventional. My thinking is that you're probably moving away from cases where `Once` is going to be clearer than the alternative of a Mutex and a bool (except for the fact that `Once` has the optimized fast path)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-:-KStalgo48zhrKjzq5n0:b-7umb4c)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Thanks for the reviews. I'll wait on your call for when you think this is safe to merge. I'm guessing later next week.

---

Review status: all files reviewed at latest revision, 22 unresolved discussions, all commit checks successful.

---

*[storage/command_queue.go, line 43 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-KRwpfHvTRDFxp3M17pG:-KStgVuTwcUeRe8wLbBd:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/494f6acdc2a5e281cfd4b38b756ee04cfb46eef4/storage/command_queue.go#L43)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> s/, pending commands/commands/
</details>
Done.

---

*[storage/command_queue.go, line 33 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-KSr4lnCLVADM2Qwwpp-:-KStgPFqGUIVi7hMQwj3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/695f5a1f988015678931bcc5291fbf3a9107d78d/storage/command_queue.go#L33)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> s/a new/the/
</details>
Done.

---

*[storage/command_queue.go, line 38 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-KSr4skiURFaqZYITGrg:-KStgSLDRyWrqlW4_vvZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/695f5a1f988015678931bcc5291fbf3a9107d78d/storage/command_queue.go#L38)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> they must also be done under the same lock.
</details>
Done.

---

*[storage/command_queue.go, line 105 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-KSr57jaSsayCuzxYyyX:-KStgbmy9vJnQT5N0477:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/695f5a1f988015678931bcc5291fbf3a9107d78d/storage/command_queue.go#L105)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> while you're here, "it" is ambiguous.
</details>
Done.

---

*[storage/command_queue.go, line 106 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-KSr5Dva9Kp4kKTNShdq:-KStgjqPKqMICyO4bFNC:b-16vl5c) ([raw file](https://github.com/cockroachdb/cockroach/blob/695f5a1f988015678931bcc5291fbf3a9107d78d/storage/command_queue.go#L106)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> add() adds the command, not the keys, right? this comment seems malformed. simply, i would remove the mentions of "add" entirely, they're covered by the god comment above.
</details>
Removed. I'll leave it to the god comment.

---

*[storage/command_queue.go, line 184 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-KSr5i6IZ0fPqT27YEPy:-KStdWrcgU7IoN4ZccMP:b-z8yi8a) ([raw file](https://github.com/cockroachdb/cockroach/blob/695f5a1f988015678931bcc5291fbf3a9107d78d/storage/command_queue.go#L184)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> can you add the word TODO somewhere? otherwise it's difficult to discern if this is done already or not
</details>
Yeah I see what you're saying: `s/can be/is/`. This is already how this all works, which is why we have the `cq.wRg.Overlaps` calls below.

---

*[storage/command_queue.go, line 338 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-KSr63MopGECURxV4It1:-KStguVJxv7NU0DBH-iF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/695f5a1f988015678931bcc5291fbf3a9107d78d/storage/command_queue.go#L338)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> just "overlapHeap is a max-heap ordered by cmd.id." anything more is asking for maintenance or rot
</details>
Done.

---

*[storage/command_queue.go, line 378 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-KSr6cRuUePmOTBTvB8t:-KSth4FlK4I9ZdCqIkQR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/695f5a1f988015678931bcc5291fbf3a9107d78d/storage/command_queue.go#L378)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> can't you just refer to the `prereqs` argument here, instead of this verbose explanation?
> 
> "add should be invoked before waiting on prereqs' pending channels."
</details>
Done.

---

*[storage/command_queue.go, line 444 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-KSr6q4G_rau-CTm9Xtj:-KSthFXOOJM6-hruf7Oo:b-mpvxlu) ([raw file](https://github.com/cockroachdb/cockroach/blob/695f5a1f988015678931bcc5291fbf3a9107d78d/storage/command_queue.go#L444)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this can happen? seems sketchy
</details>
It's an artifact of not using the command queue for inconsistent reads. I changed the logic to be more explicit.

---

*[storage/replica.go, line 1213 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-KSr8HnumoClKJkaSeBR:-KStiXJxlYDXIcssqEbT:bypifdo) ([raw file](https://github.com/cockroachdb/cockroach/blob/695f5a1f988015678931bcc5291fbf3a9107d78d/storage/replica.go#L1213)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this doesn't need to be under lock, right? in fact, it seems that only the command queue stuff needs to be under lock.
</details>
You're right. With the recent simplification this locking can reduce its scope.

---

*[storage/replica.go, line 1215 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-KSr7iHw-Jf_313SbmOH:-KStip2O-tc_5J6g7xdd:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/695f5a1f988015678931bcc5291fbf3a9107d78d/storage/replica.go#L1215)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> command's
</details>
Done.

---

*[storage/replica.go, line 1216 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-KSr7pUE_MEhQrnKfLi6:-KStjW17E2u9QVPDsp0n:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/695f5a1f988015678931bcc5291fbf3a9107d78d/storage/replica.go#L1216)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> explain why you need to truncate the capacity?
</details>
Done.

---

*[storage/replica_test.go, line 1992 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-KSr8w8qySmd7HnHygXL:-KStlVTjcq4Tfn1iEDKD:b-yikpsg) ([raw file](https://github.com/cockroachdb/cockroach/blob/695f5a1f988015678931bcc5291fbf3a9107d78d/storage/replica_test.go#L1992)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> why not use keymin and keymax for minimum arbitrariness?
</details>
For some reason that was causing the tests to time out. I'm guessing it had to do with overlapping meta keys. It seems fine to use keymax though.

---

*[storage/replica_test.go, line 2036 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-KSr98LM-hddvR0fUeoT:-KStk5u0QkZihwL4Khnm:b-667ss1) ([raw file](https://github.com/cockroachdb/cockroach/blob/695f5a1f988015678931bcc5291fbf3a9107d78d/storage/replica_test.go#L2036)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Or `rand.Perm`
</details>
And allocate for no good reason? Yuck. I'm going for highly performant tests.

---

*[storage/replica_test.go, line 2068 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-KSr9JvYktStZpfq3ePG:-KStkO8NyYFEcypjymz3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/695f5a1f988015678931bcc5291fbf3a9107d78d/storage/replica_test.go#L2068)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> unlock this before `fatal`ing to avoid deadlocks during shutdown.
</details>
Done.

---

*[util/syncutil/once.go, line 69 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-KRvHMk-JSVu5WA-8v5E-r3-69:-KStfo0nOuyilmBeMTPi:b-2u29lx) ([raw file](https://github.com/cockroachdb/cockroach/blob/c515626569cd0ea42fc82f493ba2e111acebf439/util/syncutil/once.go#L69)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> My biggest objection was just to the fact that this was not a drop-in replacement for `sync.Once`. I don't have a clear objection to adding an `IsDone` method except to observe that other similar libraries have not found a persuasive need for its addition. 
> 
> The prototypical use for `Once` is for lazy initialization of a global; creating lots of Onces and asking about their status is unconventional. My thinking is that you're probably moving away from cases where `Once` is going to be clearer than the alternative of a Mutex and a bool (except for the fact that `Once` has the optimized fast path)
</details>
Yeah it's the built-in fast path that makes `Once` seem like a desirable alternative to a Mutex and a bool for certain situations.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-:-KStkPLzW36cMeeJss7H:bucea93)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 3 of 3 files at r6, 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[storage/replica_test.go, line 1992 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-KSr8w8qySmd7HnHygXL:-KSujACE_6LGdUMWKNam:bq4m2sc) ([raw file](https://github.com/cockroachdb/cockroach/blob/695f5a1f988015678931bcc5291fbf3a9107d78d/storage/replica_test.go#L1992)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> For some reason that was causing the tests to time out. I'm guessing it had to do with overlapping meta keys. It seems fine to use keymax though.
</details>
Got it. How about using `keys.MetaMax` or `keys.SystemMax`, then?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Review status: 2 of 4 files reviewed at latest revision, 7 unresolved discussions.

---

*[storage/replica_test.go, line 1992 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-KSr8w8qySmd7HnHygXL:-KSvyguoEaZmIJQ2gzzC:bj4fq1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/695f5a1f988015678931bcc5291fbf3a9107d78d/storage/replica_test.go#L1992)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Got it. How about using `keys.MetaMax` or `keys.SystemMax`, then?
</details>
How's that?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r8, 1 of 1 files at r9.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Huh? Why'd this get closed?

tamird commented:
Because the target branch was deleted.

nvanbenschoten commented:
Is this PR ready to merge (post-rebase)? @bdarnell

tamird commented:
It's ready to go into the stability queue post-rebase.

tamird commented:
Reviewed 3 of 3 files at r10, 1 of 1 files at r11.
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Is this still a change we'd like to pursue merging? It looks like it's the only remaining stability change.

tamird commented:
If you don't mind rebasing, I don't see why not.

bdarnell commented:
I think we're still interested in getting this in, but I'm in the middle of some command queue changes for propEvalKV, so let's leave it a little longer and wait until that's in.

tamird commented:
I think this can be rebased now, though it's going to be pretty painful.

I'd suggest first `git rebase 428c30331c2ab153450afa110cba6e119e9` which is the commit that moved everything under `pkg`, and then `git rebase master  -Xignore-space-change`. Skipping the first step confuses git enough that it considers various files to have been removed without replacement.

This'll be a nice welcome back 🎁  for ya @nvanbenschoten!

nvb committed:
storage: create batchCmdSet type to hold cmds for access/scope permuations

This change also stores the different `CommandQueue` components under a
`spanScope`-indexed array.

nvb committed:
storage: add new TestCommandQueueEnclosedWrite test

This test parallels `TestCommandQueueEnclosedRead` in testing that the
command queue doesn't fail to return read/write dependencies that a
candidate read/write span depends on, despite there being an overlapping
read-only span which completely encloses the candidate.

nvanbenschoten commented:
This was rebased and updated to work with some of the new `SpanAccess` and `SpanScope` constructs, which made it easier to work with propEvalKV. @bdarnell may be interested in reviewing the first commit.

@spencerkimball may be interested in the last commit, as it expands on the testing introduced in #14733.

nvanbenschoten commented:
Thanks for the git suggestions @tamird! I was dreading the rebase across the `pkg` migration.

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> with mostly minor comments; there's one test i haven't reviewed yet because I suspect it'll be easier once it's written as a subtest. Nice work and welcome back!

---

Reviewed 7 of 7 files at r12, 3 of 4 files at r13, 1 of 1 files at r14, 1 of 1 files at r15.
Review status: 4 of 5 files reviewed at latest revision, 15 unresolved discussions, all commit checks successful.

---

*[pkg/storage/command_queue.go, line 46 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-KqZQMjE-h2HJlm6P_ip:-KqZQMjE-h2HJlm6P_iq:b93bn5o) ([raw file](https://github.com/cockroachdb/cockroach/blob/26430f84abda670d37fbaaa19bc3f7444bf7b1d5/pkg/storage/command_queue.go#L46)):*
> should

must?

---

*[pkg/storage/command_queue_test.go, line 30 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-KqZRFliRtPxuUdviJpE:-KqZRFliRtPxuUdviJpF:bnchhed) ([raw file](https://github.com/cockroachdb/cockroach/blob/26430f84abda670d37fbaaa19bc3f7444bf7b1d5/pkg/storage/command_queue_test.go#L30)):*
> ```Go
> )
> 
> var zeroTS = hlc.Timestamp{}
> ```

this is hardly used any more - can we just inline it?

---

*[pkg/storage/command_queue_test.go, line 528 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-KqZVKAbD7YFxDdaqQHS:-KqZVKAc7j22ip7qTcpG:b-r2b7ov) ([raw file](https://github.com/cockroachdb/cockroach/blob/192c1d19afb9f59d4b19e764e1d991b48c48f2c1/pkg/storage/command_queue_test.go#L528)):*
> ```Go
> //         2   2   T  ---   n/a
> // Candidate   1   F  -     spans 1, 2
> func TestCommandQueueEnclosedWrite(t *testing.T) {
> ```

lots of similarity between thse two tests as well; could they be factored into a pair of subtests?

---

*[pkg/storage/replica.go, line 2004 at r12](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-KqZH7x8SodoceYgtVTi:-KqZH7x9VZlg4zqWkUQg:b7mlu32) ([raw file](https://github.com/cockroachdb/cockroach/blob/8f05f4b5e5dac438f6c343af2d35f8b400df7dbf/pkg/storage/replica.go#L2004)):*
> ```Go
> 				return hlc.Timestamp{}
> 			}
> 			panic("unexpected scope")
> ```

```
default:
  log.Fatalf(ctx, "unexpected scope %s", scope)
```
?

---

*[pkg/storage/replica.go, line 2001 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-KqZRfNqtUsOJna0n-bQ:-KqZRfNqtUsOJna0n-bR:bh5m8yu) ([raw file](https://github.com/cockroachdb/cockroach/blob/26430f84abda670d37fbaaa19bc3f7444bf7b1d5/pkg/storage/replica.go#L2001)):*
> ```Go
> 
> 		r.cmdQMu.Lock()
> 		var prereqs [numSpanAccess][numSpanScope][]*cmd
> ```

`batchCmdSet`?

---

*[pkg/storage/replica.go, line 2027 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-KqZSQ5k7FdkTaSIqc6u:-KqZSQ5k7FdkTaSIqc6v:b-udg2c9) ([raw file](https://github.com/cockroachdb/cockroach/blob/26430f84abda670d37fbaaa19bc3f7444bf7b1d5/pkg/storage/replica.go#L2027)):*
> ```Go
> 		}
> 
> 		for i := SpanAccess(0); i < numSpanAccess; i++ {
> ```

it occurs to me that a number of these loops can be for-range loops.

---

*[pkg/storage/replica.go, line 2030 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-KqZSAKiCinhe4UnWZpo:-KqZSAKiCinhe4UnWZpp:b-u7sy23) ([raw file](https://github.com/cockroachdb/cockroach/blob/26430f84abda670d37fbaaa19bc3f7444bf7b1d5/pkg/storage/replica.go#L2030)):*
> ```Go
> 			for j := spanScope(0); j < numSpanScope; j++ {
> 				newCmd := newCmds[i][j]
> 				if newCmd == nil {
> ```

how's this possible?

---

*[pkg/storage/replica.go, line 2102 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-KqZT7ThiryYC7ZdraI2:-KqZT7ThiryYC7ZdraI3:b-bfw586) ([raw file](https://github.com/cockroachdb/cockroach/blob/26430f84abda670d37fbaaa19bc3f7444bf7b1d5/pkg/storage/replica.go#L2102)):*
> ```Go
> 
> 						// Truncate the cancelled command's prerequisite list to only those
> 						// prerequisites that have not yet been waited on. Truncate capacity to
> ```

is truncating capacity really required here? you're removing the front of the slice; is the implication that without truncating capacity the front of the slice is still somehow reachable?

---

*[pkg/storage/replica_test.go, line 2268 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-KqZUckLNecLBe992UWx:-KqZUckLNecLBe992UWy:bz66hdq) ([raw file](https://github.com/cockroachdb/cockroach/blob/26430f84abda670d37fbaaa19bc3f7444bf7b1d5/pkg/storage/replica_test.go#L2268)):*
> ```Go
> // is cancelled, and that commands dependent on cancelled commands execute
> // correctly.
> func testReplicaCommandQueueCancellation(t *testing.T, instructions []cancellationTestInstruction) {
> ```

thanks to subtests, we no longer have to write tests in this style - would it be reasonable convenient to write this as a few subtests? that would reduce the amount of jumping back and forth required to read this test.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-:-KqZVT5zaQ0twozcNz1Q:bnb0kwp)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Review status: 0 of 5 files reviewed at latest revision, 15 unresolved discussions.

---

*[pkg/storage/command_queue.go, line 46 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-KqZQMjE-h2HJlm6P_ip:-Kq_NkP3bEQcsYC00TWd:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/26430f84abda670d37fbaaa19bc3f7444bf7b1d5/pkg/storage/command_queue.go#L46)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

> should

must?
</blockquote></details>

Done.

---

*[pkg/storage/command_queue_test.go, line 30 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-KqZRFliRtPxuUdviJpE:-Kq_NqI_VK4Ov9u98CS5:b-v6tq9y) ([raw file](https://github.com/cockroachdb/cockroach/blob/26430f84abda670d37fbaaa19bc3f7444bf7b1d5/pkg/storage/command_queue_test.go#L30)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this is hardly used any more - can we just inline it?
</blockquote></details>

It's still used 20 times, so I think it's still worth keeping.

---

*[pkg/storage/command_queue_test.go, line 528 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-KqZVKAbD7YFxDdaqQHS:-KqcsWKOH2owTKQjbU73:b-eroaaq) ([raw file](https://github.com/cockroachdb/cockroach/blob/192c1d19afb9f59d4b19e764e1d991b48c48f2c1/pkg/storage/command_queue_test.go#L528)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

lots of similarity between thse two tests as well; could they be factored into a pair of subtests?
</blockquote></details>

These are both testing very specific scenarios, so in this case I don't want to obscure that by pulling them into subtests and being forced to generalize what they're testing.

---

*[pkg/storage/replica.go, line 2004 at r12](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-KqZH7x8SodoceYgtVTi:-Kq_OI1uvhehME2G4Nzi:besfuj0) ([raw file](https://github.com/cockroachdb/cockroach/blob/8f05f4b5e5dac438f6c343af2d35f8b400df7dbf/pkg/storage/replica.go#L2004)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

```
default:
  log.Fatalf(ctx, "unexpected scope %s", scope)
```
?
</blockquote></details>

Go still hasn't gotten smart enough to detect these cases without shouting about a `missing return at end of function`. 

---

*[pkg/storage/replica.go, line 2001 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-KqZRfNqtUsOJna0n-bQ:-Kq_P10AR5fGT9t6WKZz:bz1msqn) ([raw file](https://github.com/cockroachdb/cockroach/blob/26430f84abda670d37fbaaa19bc3f7444bf7b1d5/pkg/storage/replica.go#L2001)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

`batchCmdSet`?
</blockquote></details>

I wish, but this holds a slice of prereq `*cmd`s for each access/scope combination. `[numSpanAccess][numSpanScope]*cmd` vs `[numSpanAccess][numSpanScope][]*cmd`.

---

*[pkg/storage/replica.go, line 2027 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-KqZSQ5k7FdkTaSIqc6u:-Kq_PRkhUL6pFjCiBokI:bxgp82x) ([raw file](https://github.com/cockroachdb/cockroach/blob/26430f84abda670d37fbaaa19bc3f7444bf7b1d5/pkg/storage/replica.go#L2027)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

it occurs to me that a number of these loops can be for-range loops.
</blockquote></details>

Done where possible. Some of these result in the need for a lot of `int->SpanAccess/spanScope` casts because of how `for-range` indexing works.

---

*[pkg/storage/replica.go, line 2030 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-KqZSAKiCinhe4UnWZpo:-Kq_SHKoicN_O49DXo79:boag87p) ([raw file](https://github.com/cockroachdb/cockroach/blob/26430f84abda670d37fbaaa19bc3f7444bf7b1d5/pkg/storage/replica.go#L2030)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

how's this possible?
</blockquote></details>

Commented.

---

*[pkg/storage/replica.go, line 2102 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-KqZT7ThiryYC7ZdraI2:-Kq_SobLruaTOEeJlTmA:bpqtcp3) ([raw file](https://github.com/cockroachdb/cockroach/blob/26430f84abda670d37fbaaa19bc3f7444bf7b1d5/pkg/storage/replica.go#L2102)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

is truncating capacity really required here? you're removing the front of the slice; is the implication that without truncating capacity the front of the slice is still somehow reachable?
</blockquote></details>

No I don't think it's reachable, but I seem to remember truncation like this allowing the front of the slice to be GCed. This could be important if it means we're preventing a big chunk of the dependency graph from being garbage collected, especially in a cascading cancellation situation. I could very well be wrong though.

---

*[pkg/storage/replica_test.go, line 2268 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-KqZUckLNecLBe992UWx:-KqcsQAALJV4rn0I3Mxv:bdmesjw) ([raw file](https://github.com/cockroachdb/cockroach/blob/26430f84abda670d37fbaaa19bc3f7444bf7b1d5/pkg/storage/replica_test.go#L2268)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

thanks to subtests, we no longer have to write tests in this style - would it be reasonable convenient to write this as a few subtests? that would reduce the amount of jumping back and forth required to read this test.
</blockquote></details>

Done. Used as an excuse to add a lot more tests too.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 4 files at r13, 1 of 5 files at r16, 2 of 4 files at r17, 1 of 1 files at r18, 1 of 1 files at r19.
Review status: all files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica.go, line 2004 at r12](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-KqZH7x8SodoceYgtVTi:-Kqd37G5mD0VDTWQ37Db:b-nm0sps) ([raw file](https://github.com/cockroachdb/cockroach/blob/8f05f4b5e5dac438f6c343af2d35f8b400df7dbf/pkg/storage/replica.go#L2004)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Go still hasn't gotten smart enough to detect these cases without shouting about a `missing return at end of function`. 
</blockquote></details>

Ah, OK. Still, we should at least include something useful in the message.

---

*[pkg/storage/replica.go, line 2102 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-KqZT7ThiryYC7ZdraI2:-Kqd3SrSQ50xwLG8oOrL:b-basqpa) ([raw file](https://github.com/cockroachdb/cockroach/blob/26430f84abda670d37fbaaa19bc3f7444bf7b1d5/pkg/storage/replica.go#L2102)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

No I don't think it's reachable, but I seem to remember truncation like this allowing the front of the slice to be GCed. This could be important if it means we're preventing a big chunk of the dependency graph from being garbage collected, especially in a cascading cancellation situation. I could very well be wrong though.
</blockquote></details>

Yeah, I agree that it's important - can you verify that it does what we think and comment with a reference?

---

*[pkg/storage/replica_test.go, line 2381 at r19](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-Kqd4ppIGTIV1Rki9On3:-Kqd4ppIGTIV1Rki9On4:b-mcgpwi) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb665bda392225fd7a910eff2ea24df693819054/pkg/storage/replica_test.go#L2381)):*
> verifies that commands in a random
> // which dependency graph 

I think "which" and "dependency graph" need to switch places ?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Review status: 1 of 5 files reviewed at latest revision, 9 unresolved discussions.

---

*[pkg/storage/replica.go, line 2004 at r12](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-KqZH7x8SodoceYgtVTi:-KqddGqnTI8hRxoDpG_n:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8f05f4b5e5dac438f6c343af2d35f8b400df7dbf/pkg/storage/replica.go#L2004)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Ah, OK. Still, we should at least include something useful in the message.
</blockquote></details>

Done.

---

*[pkg/storage/replica.go, line 2102 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-KqZT7ThiryYC7ZdraI2:-KqdiI8a7-3kYdKZ1Enr:bd03bfb) ([raw file](https://github.com/cockroachdb/cockroach/blob/26430f84abda670d37fbaaa19bc3f7444bf7b1d5/pkg/storage/replica.go#L2102)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Yeah, I agree that it's important - can you verify that it does what we think and comment with a reference?
</blockquote></details>

I don't know why I was under the impression that that would work, but I couldn't find any references to support my claim. For now, I'll just nil out the prefix of the slice.

---

*[pkg/storage/replica_test.go, line 2381 at r19](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-Kqd4ppIGTIV1Rki9On3:-KqddUTAmAdHwwWwVkLp:b-1q0bc2) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb665bda392225fd7a910eff2ea24df693819054/pkg/storage/replica_test.go#L2381)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

> verifies that commands in a random
> // which dependency graph 

I think "which" and "dependency graph" need to switch places ?
</blockquote></details>

Done. Can't slip anything by you.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 4 files at r21, 1 of 1 files at r22, 1 of 1 files at r23.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[pkg/storage/replica.go, line 2107 at r23](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-KqdjhqnUAmHQOEhQ9Ts:-KqdjhqnUAmHQOEhQ9Tt:b-mhqixm) ([raw file](https://github.com/cockroachdb/cockroach/blob/9eea39ba1837ffc78d00dfce00874dc56b2ef273/pkg/storage/replica.go#L2107)):*
> ```Go
> 						// that have already completed. This prevents us from leaking large
> 						// chunks of the dependency graph.
> 						prs := len(newCmd.prereqs)
> ```

`prs` is no longer used, the below can be `newCmd.prereqs = newCmd.prereqs[i:]`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[pkg/storage/replica.go, line 2107 at r23](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448#-KqdjhqnUAmHQOEhQ9Ts:-Kqdk7VjE9iAITtSBFHy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9eea39ba1837ffc78d00dfce00874dc56b2ef273/pkg/storage/replica.go#L2107)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

`prs` is no longer used, the below can be `newCmd.prereqs = newCmd.prereqs[i:]`
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 3 files at r24, 1 of 1 files at r25, 1 of 1 files at r26.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9448)*
<!-- Sent from Reviewable.io -->
